Docs: Update query and resource caching documentation to improve clarity and add additional context (#70556)
grafana/grafana: docs/sources/administration/data-source-management/index.md

  - ../datasources/datasource_permissions/
  - ../enterprise/datasource_permissions/
  - ../enterprise/query-caching/

>   - ../enterprise/query-and-resource-caching/

Reviewer: Did this page used to exist? Just checking that we in fact need an alias here.

Author: Ah okay - I didn't realize that's what aliases were for. I will remove it.

Author: Done [d2fdcf7](https://github.com/grafana/grafana/pull/70556/commits/d2fdcf7a9fa2854e04329ff81095b316ab8e5e6b)

A "query" refers to a data source query for data frames to be transformed and/or visualized in Grafana. A "resource" refers to any HTTP requests a plugin might make, for example: the Amazon Timestream plugin requesting a list of available databases from AWS. For more information on data source queries and resources, please see the developers page on [backend plugins]({{< relref "../../developers/plugins/backend/" >}}).

The caching feature works for **all** backend data sources. You can enable the cache globally in Grafana's [configuration]({{< relref "../../setup-grafana/configure-grafana/enterprise-configuration/#caching" >}}), and configure a cache duration (also called Time to Live, or TTL) for each data source individually.

{{% admonition type="note" %}}
Available in [Grafana Enterprise]({{< relref "../../introduction/grafana-enterprise/" >}}) and [Grafana Cloud Pro and Advanced](/docs/grafana-cloud/).
{{% /admonition %}}

> The following cache backend integrations are available: in-memory, Redis, and Memcached.

Reviewer 2: What's the reasoning behind adding `integrations` here?

Author: no strong reason - I just thought it sounded a bit more natural. I could also say "backend implementations" or just "backend options" if one of those seems better.

Reviewer 2: Oh I like `backend options`, if that works for you!

Author: Done [64f1100](https://github.com/grafana/grafana/pull/70556/commits/64f1100386bdd1390bf01a09ab9e6175f9fb9beb)

{{% admonition type="note" %}}
Storing cached queries in-memory can increase Grafana's memory footprint. In production environments, a Redis or Memcached backend is highly recommended.
{{% /admonition %}}

> When a dashboard panel queries a data source with cached data, whether the query fetches fresh data or uses cached data is determined by the panel's **interval.** The Grafana backend uses the interval to round the query time range to a nearby cached time range, increasing the likelihood of cache hits. Therefore, wider panels and dashboards with shorter time ranges fetch new data more frequently than narrower panels and dashboards with longer time ranges.

Reviewer 2: Does it have to be a dashboard? If I remember correctly, explore also uses the cache.

Author: good point, reading it back I don't think adding "dashboard" was necessary at all.

> When a dashboard panel queries a data source with cached data, whether the query fetches fresh data or uses cached data is determined by the panel's **interval.** The Grafana backend uses the interval to round the query time range to a nearby cached time range, increasing the likelihood of cache hits. Therefore, wider panels and dashboards with shorter time ranges fetch new data more frequently than narrower panels and dashboards with longer time ranges.

Reviewer 2: What do you think about this?
```suggestion
When a panel queries a data source with cached data, it will either fetch fresh data or use cached data based on  the panel's **interval.** The interval is used to round the query time range to a nearby cached time range, increasing the likelihood of cache hits. Therefore, wider panels and dashboards with shorter time ranges fetch new data more frequently than narrower panels and dashboards with longer time ranges.
```

Author: looks great!

<div class="clearfix"></div>

## Query and resource caching

> When query and resource caching is enabled, Grafana temporarily stores the results of data source queries and resource requests. When you or another user submit the exact same query or resource request again, the results will come back from the cache instead of from the data source itself.

Reviewer: ```suggestion
When you enable query and resource caching, Grafana temporarily stores the results of data source queries and resource requests. When you or another user submit the same query or resource request again, the results will come back from the cache instead of from the data source.
```

> A "query" refers to a data source query for data frames to be transformed and/or visualized in Grafana. A "resource" refers to any HTTP requests a plugin might make, for example: the Amazon Timestream plugin requesting a list of available databases from AWS. For more information on data source queries and resources, please see the developers page on [backend plugins]({{< relref "../../developers/plugins/backend/" >}}).

Reviewer: ```suggestion
When using Grafana, a query pertains to a request for data frames to be modified or displayed. A resource relates to any HTTP requests made by a plugin, such as the Amazon Timestream plugin requesting a list of available databases from AWS. For more information on data source queries and resources, please see the developers page on [backend plugins]({{< relref "../../developers/plugins/backend/" >}}).
```

{{% admonition type="note" %}}
Available in [Grafana Enterprise]({{< relref "../../introduction/grafana-enterprise/" >}}) and [Grafana Cloud Pro and Advanced](/docs/grafana-cloud/).
{{% /admonition %}}

The following cache backend integrations are available: in-memory, Redis, and Memcached.

{{% admonition type="note" %}}
Storing cached queries in-memory can increase Grafana's memory footprint. In production environments, a Redis or Memcached backend is highly recommended.
{{% /admonition %}}

When a panel queries a data source with cached data, it will either fetch fresh data or use cached data depending on the panel's **interval.** The interval is used to round the query time range to a nearby cached time range, increasing the likelihood of cache hits. Therefore, wider panels and dashboards with shorter time ranges fetch new data more often than narrower panels and dashboards with longer time ranges.

> A panel's interval is visible in the [query options]({{< relref "../../panels-visualizations/query-transform-data/" >}}). It is calculated like this: `time range / max data points`. Max data points are calculated based on the width of the panel. For example, a wide panel with `1000 data points` on a dashboard with a time range of `last 7 days` will retrieve fresh data every 10 minutes: `7d / 1000 = 10m`. In this example, cached data for this panel will be served for up to 10 minutes before Grafana needs to query the data source again for new data.

Reviewer: ```suggestion
A panel's interval is visible in the [query options]({{< relref "../../panels-visualizations/query-transform-data/" >}}). It is calculated as follows: `time range / max data points`. Max data points are calculated based on the width of the panel. For example, a wide panel with `1000 data points` on a dashboard with a time range of `last 7 days` will retrieve fresh data every 10 minutes: `7d / 1000 = 10m`. In this example, cached data for this panel will be served for up to 10 minutes before Grafana needs to query the data source again for new data.
```

> You can make a panel retrieve fresh data more often by increasing the **Max data points** setting in the panel's [query options]({{< relref "../../panels-visualizations/query-transform-data/" >}}).

Reviewer: ```suggestion
You can configure a panel to retrieve data more often by increasing the **Max data points** setting in the panel's [query options]({{< relref "../../panels-visualizations/query-transform-data/" >}}).
```

### Caching benefits

By reducing the number of queries and requests sent to data sources, caching can provide the following benefits:

- Faster dashboard load times, especially for popular dashboards.
- Reduced API costs.
- Reduced likelihood that APIs will rate-limit or throttle requests.

### Data sources that work with query caching

> Query caching works for all backend data sources. More specifically, caching works with Grafana's [built-in data sources]({{< relref "../../datasources/#built-in-core-data-sources" >}}), and any [backend data source plugins](https://grafana.com/grafana/plugins/?type=datasource) that extend the `DataSourceWithBackend` class in the plugins SDK.

Reviewer: ```suggestion
Query caching works for Grafana's [built-in data sources]({{< relref "../../datasources/#built-in-core-data-sources" >}}), and [backend data source plugins](https://grafana.com/grafana/plugins/?type=datasource) that extend the `DataSourceWithBackend` class in the plugins SDK.
```

> To verify that a data source works with query caching, follow the instructions below to **Enable and Configure query caching**. If caching is enabled in Grafana but the Caching tab is not visible for the given data source, then query caching is not available for that data source.

Reviewer: Add a link to the instructions below.

Author: Done on [3b46638](https://github.com/grafana/grafana/pull/70556/commits/3b46638c9afcdda4905bb4941b9d069440705565)

### Data sources that work with query caching

Query caching works for all backend data sources. More specifically, caching works with Grafana's [built-in data sources]({{< relref "../../datasources/#built-in-core-data-sources" >}}), and any [backend data source plugins](https://grafana.com/grafana/plugins/?type=datasource) that extend the `DataSourceWithBackend` class in the plugins SDK.

To verify that a data source works with query caching, follow the instructions below to **Enable and Configure query caching**. If caching is enabled in Grafana but the Caching tab is not visible for the given data source, then query caching is not available for that data source.

{{% admonition type="note" %}}

> Some data sources, such as Elasticsearch, Prometheus, and Loki, cache queries themselves, so Grafana query caching does not significantly improve performance.

Reviewer 2: I'm not sure if this is too niche to include in the docs, but Loki (and perhaps the others) do not cache resource requests, so adding resource caching can be really helpful if they have large variable queries on panels.

Author: Maybe we could find a way to say this in the #caching-benefits section? We could say something about how resource caching avoids unnecessary duplicate requests for static data, and maybe give an example about Loki. Feel free to take a stab at it

Author: Thinking more about this - we now link to the backend plugins page which covers the use cases for resource caching, and the benefits of resource caching are covered in the benefits section. I think we are probably okay leaving out the loki detail. WDYT?

Reviewer 2: Ah good point! Maybe we could point it out here like this?
```
Some data sources, such as Elasticsearch, Prometheus, and Loki, cache queries themselves, so Grafana **query** caching does not significantly improve performance, though **resource** caching may help. See [these docs](<insert backend plugins page>) for details.
```
There has been some confusion in the past on this with some larger customers, so I could see it being useful, even if it's a bit redundant. Wdyt?

Author: I like it, how is this: [e61bc48](https://github.com/grafana/grafana/pull/70556/commits/e61bc48466d1c372e644d6cf04182a30bc495da3)
